### ISSUE TITLE
Add `User-Agent` in link checking

### ISSUE THREAD

**Issue by andrewvaughan** [oxsecurity/megalinter#3304]
**Is your feature request related to a problem? Please describe.**
Many websites block (`403` response) requests without a `User-Agent` HTTP request header set. This causes link checkers to automatically fail.

For https://github.com/tcort/markdown-link-check/ a proper issue has already been raised (https://github.com/tcort/markdown-link-check/issues/172); however, given that, for one, this issue is now almost 3-years old without a response, and, for two, it's better for each individual client to provide their unique `User-Agent` to be a good netizen, I recommend having MegaLinter provide a versioned `User-Agent` in their default configurations.

**Describe the solution you'd like**
Add the following to the default https://github.com/oxsecurity/megalinter/blob/main/TEMPLATES/.markdown-link-check.json configuration, but also to any other link-checking linters that may exist:

```json
{
  // ... existing config ...

  "httpHeaders": [
    {
      "urls": ["http", ".", "/"],
      "headers": {
        "User-Agent": "Mozilla/5.0 (compatible; markdown-link-check/3.11.2; MegaLinter/7.8.0; +https://megalinter.io)"
      }
    }
  ]
```

For more information on `User-Agent` header best practices and why I recommend the above:
<https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/User-Agent>

**Describe alternatives you've considered**
- Add places like `stackoverflow.com` to my ignore list.
- Cry

**Additional context**

My MegaLinter right now:

![Please?](https://media0.giphy.com/media/yx400dIdkwWdsCgWYp/giphy.gif)

**Comment by andrewvaughan** [comment]
So while I still think this is a good idea, I'm still digging further, because while this resolves me running `npx markdown-link-check` directly on my machine, if I run this command on the `megalinter` container, it still fails...

```sh
docker exec -it megalinter markdown-link-check -c /tmp/lint/.config/linters/.markdown-link-check.json -p /tmp/lint/.github/SUPPORT.md
```

**Comment by andrewvaughan** [comment]
More progress - and I think this actually may need to be either moved or duplicated to https://github.com/tcort/markdown-link-check now, because of what I found:

```sh
docker run -v ${PWD}:/tmp:ro --rm -i ghcr.io/tcort/markdown-link-check:stable -c /tmp/.config/linters/.markdown-link-check.json /tmp/.github/SUPPORT.md

# ... the 403 errors re-appear
```

It seems that the issue is the user-agent (or _something_) is not working, but only while run within a Docker container (or something specific about how either both the Megalinter and this Docker container are built/configured).

I'm going to do some more digging to see if it's an issue with a commonality, Docker, or otherwise, but this is a deeper issue than I expected.

**Comment by andrewvaughan** [comment]
More updates... _it works fine with `wget`_... well, at least on Alpine. Nothing else. wat.

```sh
# Local works fine...
$ wget --server-response -U "Mozilla/5.0 (compatible; markdown-link-check/3.11.2; MegaLinter/7.7.0; +https://megalinter.io)" https://meta.stackexchange.com/questions/66377/what-is-the-xy-problem/66378#66378
--2024-01-21 16:55:55--  https://meta.stackexchange.com/questions/66377/what-is-the-xy-problem/66378
Resolving meta.stackexchange.com (meta.stackexchange.com)... 172.64.144.30, 104.18.43.226
Connecting to meta.stackexchange.com (meta.stackexchange.com)|172.64.144.30|:443... connected.
HTTP request sent, awaiting response... 
  HTTP/1.1 200 OK

# Alpine works fine...
$ docker run -it --rm alpine sh -c 'apk update -q; apk add -q wget; wget --server-response -U "Mozilla/5.0 (compatible; markdown-link-check/3.11.2; MegaLinter/7.7.0; +https://megalinter.io)" https://meta.stackexchange.com/questions/66377/what-is-the-xy-problem/66378#66378'
--2024-01-21 22:03:56--  https://meta.stackexchange.com/questions/66377/what-is-the-xy-problem/66378
Resolving meta.stackexchange.com (meta.stackexchange.com)... 172.64.144.30, 104.18.43.226
Connecting to meta.stackexchange.com (meta.stackexchange.com)|172.64.144.30|:443... connected.
HTTP request sent, awaiting response... 
  HTTP/1.1 200 OK

# Ubuntu works fine with wget
$ docker run -it --rm --entrypoint /bin/sh ubuntu -c 'apt update -qq; apt install -y -qq wget; wget --server-response -U "Mozilla/5.0 (compatible; markdown-link-check/3.11.2; MegaLinter/7.7.0; +https://megalinter.io)" https://meta.stackexchange.com/questions/66377/what-is-the-xy-problem/66378#66378'

# ...

Connecting to meta.stackexchange.com (meta.stackexchange.com)|104.18.43.226|:443... connected.
HTTP request sent, awaiting response... 
  HTTP/1.1 200 OK

# But megalinter, not so much...
$ docker run --entrypoint /bin/bash -it --rm oxsecurity/megalinter-python:v7.7.0 -c 'wget --server-response -U "Mozilla/5.0 (compatible; markdown-link-check/3.11.2; MegaLinter/7.7.0; +https://megalinter.io)" https://meta.stackexchange.com/questions/66377/what-is-the-xy-problem/66378#66378'
Connecting to meta.stackexchange.com (104.18.43.226:443)
  HTTP/1.1 403 Forbidden
wget: server returned error: HTTP/1.1 403 Forbidden
```

**Comment by nvuillam** [comment]
What an investigation @andrewvaughan  :D 

it seems markdown-links-check
- relies on link-check https://www.npmjs.com/package/link-check
- who itself relies on needle https://www.npmjs.com/package/needle

Maybe needle has different behaviors depending of environment variables ?

Something to check that could be to expose a mock service, and log the calls within docker and out of docker to see the differences :)

**Comment by andrewvaughan** [comment]
> What an investigation @andrewvaughan :D
> 
> it seems markdown-links-check
> 
> * relies on link-check [npmjs.com/package/link-check](https://www.npmjs.com/package/link-check)
> * who itself relies on needle [npmjs.com/package/needle](https://www.npmjs.com/package/needle)
> 
> Maybe needle has different behaviors depending of environment variables ?
> 
> Something to check that could be to expose a mock service, and log the calls within docker and out of docker to see the differences :)

Lol you should see the comment I was half-way through writing...

I have gone the depths of the dependency stack. I am weary and tired, but I bear the fruits of my labor:

* https://github.com/oxsecurity/megalinter
  * Depends on https://github.com/tcort/markdown-link-check for markdown link validation
    * Depends on https://github.com/tcort/link-check for agnostic link validation
      * [Depends](https://github.com/tcort/link-check/blob/master/lib/proto/http.js#L56) on https://github.com/tomas/needle for making `http(s)` calls

Bear with me friends, because this is where my soul started tearing apart. The code was a jungle.

* I broke my soul finding the [needle.head() export](https://github.com/tomas/needle/blob/9454d7bdc94b5a9a9c8f60eafefb6408f83a9a37/lib/needle.js#L844)
  * Sans a comment in sight, I [found the request call](https://github.com/tomas/needle/blob/9454d7bdc94b5a9a9c8f60eafefb6408f83a9a37/lib/needle.js#L387)
    * I found they were tossing the [hard work](https://github.com/tomas/needle/blob/9454d7bdc94b5a9a9c8f60eafefb6408f83a9a37/lib/needle.js#L455) off to the built-in Node.js [https](https://nodejs.org/api/https.html) library when actually [making the request](https://github.com/tomas/needle/blob/9454d7bdc94b5a9a9c8f60eafefb6408f83a9a37/lib/needle.js#L512)

Which brought me to the Node.js [core source-code](https://github.com/nodejs/node/blob/v21.6.0/lib/https.js#L361) with even LESS documentation...

![Ergo, visa-vis](https://s.yimg.com/ny/api/res/1.2/B0mcbBtIXSFlnW.7u4Jm2w--/YXBwaWQ9aGlnaGxhbmRlcjt3PTY0MDtoPTM1Mg--/https://media.zenfs.com/en/spy_453/d9b36cc1e1a1ed70caf4ead9e2c7ff17)

* The [ClientRequest](https://github.com/nodejs/node/blob/main/lib/_http_client.js#L130)
  * With no Agent set, the [undici](https://github.com/nodejs/undici) library is used and [initiates a connection](https://github.com/nodejs/node/blob/main/lib/_http_client.js#L334)
    * https://github.com/nodejs/node/blob/main/lib/https.js#L138
      * https://github.com/nodejs/node/blob/main/lib/_tls_wrap.js#L1739
        * https://github.com/nodejs/node/blob/main/lib/_tls_wrap.js#L512
          * 

...that's about as far as I got

**Comment by andrewvaughan** [comment]
Narrowed it down:

```sh
$ docker run --entrypoint /bin/bash -it --rm oxsecurity/megalinter-python:v7.7.0

# curl --no-alpn -I -A "Mozilla/5.0 (compatible; markdown-link-check/3.11.2; MegaLinter/7.7.0; +https://megalinter.io)" "https://meta.stackexchange.com/questions/66377/what-is-the-xy-problem/66378#66378"
HTTP/1.1 200 OK

# curl -I -A "Mozilla/5.0 (compatible; markdown-link-check/3.11.2; MegaLinter/7.7.0; +https://megalinter.io)" "https://meta.stackexchange.com/questions/66377/what-is-the-xy-problem/66378#66378"
HTTP/2 403
```

There must be something either about the `HTTP/2` protocol or [ALPN](https://en.wikipedia.org/wiki/Application-Layer_Protocol_Negotiation) that is flagging the linter as an unwanted bot. My guess is the GNU `wget` implementations enforce HTTP/1.1

Now **why** these remote hosts are allowing non-ALPN traffic through but blocking ALPN traffic is an interesting question.

I will move the remainder of this discovery over to an issue on the markdown-link-checker, but I would definitely consider adding a unique UserAgent to Megalinter with this Issue - it will help prevent the default `Needle/X.X` UserAgent that gets applied from getting over-blocked.

**Comment by echoix** [comment]
Wow, what a pleasure to read at the whole thinking process.

Near the beginning of the thread, I was thinking on trying a Debian/debian slim/ubuntu container too. Sometimes, to make sure that I'm not hitting some particular differences of musl-based packages, it's always good to check if it should work without it. In the last couple of years of still being subscribed to notifications on the node-red docker repo, you'd be surprised about the frequency of weird behaviours that doesn't happen with a Debian based base image (as they have both).


I'd never thought of going as deep as you did, you even learned me a new word, ALPN!
Your links to the source code in node.js, in a well written debugging summary like this would be good candidates to being permalinks to be able to reread the good thing in a months time (the issues that you made that would be referencing this might take a while).


As for the user agent, I have three contradicting opinions. On one side, it is reasonable and your explanations justify correctly the need to have a user agent. On another side, shouldn't it be a user agent for the linter rather than Megalinter? While you are talking specifically talking about markdown-link-checker, the linter I struggle a bit more with lychee. That brings to the third competing opinion: some sites answer completely differently by user-agent. _Wink Wink SourceForge_. Even though I already found it out on myself before, it was apparent when working with a winget definition for a new software version, where the download URLs work only in specific cases. (Luckily they have an arrangement so their CI works better than locally). But these differences came back at the beginning of the introduction of lychee linter, before getting stuff smoothed out. So here, sometimes having the generic most common user agent is the only way to have a (badly) configured website to work at all.

So I can't decide yet what will weight more in the balance.

**Comment by andrewvaughan** [comment]
Thanks for the kind words!

Per your concerns on the UA - you're 100% on point. That's why I particularly recommended the pattern of UA that I did. There's a link I put above with best-practices on generating UAs.

Most "crawlers" literally put "crawler/2.2.2" which can be problematic, if only because some lazy admins block "everything not standard," which was never intended for UAs. That's where marking a `compatible` comes in.

The UA format is `<name>/<version> <comment>` - very, very simple. Filtering is only _supposed_ to happen on the first two, but platform standard needs did add some filtering in the comment area informally.

As such, nearly all UAs for browsers are:

`BrowserName/1.1 (System Information)`

With some level of standardization in what the `(System Information)` entails.

However! That comment can technically be anything - and there actually is a better pattern for applications that meet the "requirements" of a browser standard but make use of it in a different way; for example:

`Mozilla/5.0 (compatible; technology/x.x.x; technology/x.x.x; +https://reference)`

This is a _great_ pattern, because it both informs the server as to what standard can be managed _and_ allows for fine-tune bot management by administrators. Maybe someone wants to block all of MegaLinter - maybe just link checkers. Maybe just particular, problematic versions. It's their choice in this format with some simple string-matching.

So you end up with something like the recommendation above, or, for something more simple, the following:

`Mozilla/5.0 (compatible; MegaLinter/7.8.0; +https://megalinter.io)`

The reference URL at the end is also super helpful - coming from an admin, if I were to start seeing this new UA appear out of everywhere, my first reaction would be to block it. A responsible admin, however, will check the reference to see what its purpose is and determine as to whether it is nefarious or not for the purposes of the applications. Systems like Web Application Firewalls learn from this, and you might even start to see `MegaLinter` UAs have fewer and fewer issues with systems like Cloudflare WAF (which ended up being the primary cause of the problem, here).

Unfortunately, without any specification, you end up with the default for whatever the linters are - or sometimes no UA at all. For `markdown-link-check`, I believe it's using the default from the `needle` library it's using, which is `needle/x.x.x`.

Now, imagine how many people have probably used the `needle` dependency to make more ...problematic... bots for servers. How could they tell MegaLinter traffic from those bad bots? There's really no way. Easier to just block the entire batch, and you'll be much more likely to get picked up in a WAF as a nefarious bot.

So, for me - I think the question is whether the responsibility of setting an appropriate UA is for the tool or the tool container. I lean toward the argument that the UA should always represent the technology closest to the end user (in this case, MegaLinter, being the utility I chose to incorporate into my project, not necessarily the specific linter), so I would prefer my UA to represent `Mozilla/5.0 (compatible; MegaLinter/7.8.0; +https://megalinter.io)` - but this is a personal opinion. There's a strong argument to include the individual linter details in each UA, as well.

This is just me thinking out loud, but that's my $0.00002 on the issue!
